Description: Don't truncate process title in syslogs
 setproctile() tweakery alters memory buffer referenced by __progname, which
 make the logs look bad.  Set __progname to a safe static buffer early on to
 make everybody happy.  While there, check for overflow to make gilles happy.
Author: Eric Faurot <eric@faurot.net>
Origin: Upstream
Applied-Upstream: https://github.com/poolpOrg/OpenSMTPD/commit/1bf55e27
Bug: https://github.com/poolpOrg/OpenSMTPD/issues/239
Bug-Debian: http://bugs.debian.org/724062
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
Index: opensmtpd/openbsd-compat/setproctitle.c
===================================================================
--- opensmtpd.orig/openbsd-compat/setproctitle.c	2013-10-20 07:54:16.093643110 -0400
+++ opensmtpd/openbsd-compat/setproctitle.c	2013-10-20 08:05:08.059495629 -0400
@@ -64,6 +64,9 @@
 
 #endif /* HAVE_SETPROCTITLE */
 
+static char altprogname[1024];
+extern char *__progname;
+
 void
 compat_init_setproctitle(int argc, char *argv[])
 {
@@ -73,6 +76,9 @@
 	char **envp = environ;
 	int i;
 
+	strlcpy(altprogname, __progname, sizeof(altprogname));
+	__progname = altprogname;
+
 	/*
 	 * NB: This assumes that argv has already been copied out of the
 	 * way. This is true for sshd, but may not be true for other 
@@ -125,7 +131,7 @@
 	va_list ap;
 	char buf[1024], ptitle[1024];
 	size_t len;
-	extern char *__progname;
+	int r;
 #if SPT_TYPE == SPT_PSTAT
 	union pstun pst;
 #endif
@@ -137,13 +143,16 @@
 
 	strlcpy(buf, __progname, sizeof(buf));
 
+	r = -1;
 	va_start(ap, fmt);
 	if (fmt != NULL) {
 		len = strlcat(buf, ": ", sizeof(buf));
 		if (len < sizeof(buf))
-			vsnprintf(buf + len, sizeof(buf) - len , fmt, ap);
+			r = vsnprintf(buf + len, sizeof(buf) - len , fmt, ap);
 	}
 	va_end(ap);
+	if (r == -1 || (size_t)r >= sizeof(buf) - len)
+		return;
 	strnvis(ptitle, buf, sizeof(ptitle),
 	    VIS_CSTYLE|VIS_NL|VIS_TAB|VIS_OCTAL);
 
